// tdesign-vue-next/cjs/link/props.js
/**
 * tdesign v0.24.9
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  content: {
    type: [String, Function]
  },
  "default": {
    type: [String, Function]
  },
  disabled: Boolean,
  hover: {
    type: String,
    "default": "underline",
    validator: function validator(val) {
      if (!val) return true;
      return ["color", "underline"].includes(val);
    }
  },
  href: {
    type: String,
    "default": ""
  },
  prefixIcon: {
    type: Function
  },
  size: {
    type: String,
    validator: function validator(val) {
      if (!val) return true;
      return ["small", "medium", "large"].includes(val);
    }
  },
  suffixIcon: {
    type: Function
  },
  target: {
    type: String,
    "default": ""
  },
  theme: {
    type: String,
    "default": "default",
    validator: function validator(val) {
      if (!val) return true;
      return ["default", "primary", "danger", "warning", "success"].includes(val);
    }
  },
  underline: Boolean,
  onClick: Function
};

exports["default"] = props;
