// tdesign-vue-next/cjs/time-picker/time-range-picker-props.js
/**
 * tdesign v0.24.9
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  allowInput: Boolean,
  clearable: Boolean,
  disableTime: {
    type: Function
  },
  disabled: {
    type: [Boolean, Array],
    "default": false
  },
  format: {
    type: String,
    "default": "HH:mm:ss"
  },
  hideDisabledTime: {
    type: Boolean,
    "default": true
  },
  placeholder: {
    type: [String, Array],
    "default": void 0
  },
  popupProps: {
    type: Object
  },
  presets: {
    type: Object
  },
  rangeInputProps: {
    type: Object
  },
  size: {
    type: String,
    "default": "medium",
    validator: function validator(val) {
      if (!val) return true;
      return ["small", "medium", "large"].includes(val);
    }
  },
  steps: {
    type: Array,
    "default": function _default() {
      return [1, 1, 1];
    }
  },
  value: {
    type: Array,
    "default": void 0
  },
  modelValue: {
    type: Array,
    "default": void 0
  },
  defaultValue: {
    type: Array
  },
  onBlur: Function,
  onChange: Function,
  onFocus: Function,
  onInput: Function,
  onPick: Function
};

exports["default"] = props;
