// tdesign-vue-next/cjs/tabs/props.js
/**
 * tdesign v0.24.9
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  addable: Boolean,
  disabled: Boolean,
  dragSort: Boolean,
  list: {
    type: Array
  },
  placement: {
    type: String,
    "default": "top",
    validator: function validator(val) {
      if (!val) return true;
      return ["left", "top", "bottom", "right"].includes(val);
    }
  },
  size: {
    type: String,
    "default": "medium",
    validator: function validator(val) {
      if (!val) return true;
      return ["medium", "large"].includes(val);
    }
  },
  theme: {
    type: String,
    "default": "normal",
    validator: function validator(val) {
      if (!val) return true;
      return ["normal", "card"].includes(val);
    }
  },
  value: {
    type: [String, Number],
    "default": void 0
  },
  modelValue: {
    type: [String, Number],
    "default": void 0
  },
  defaultValue: {
    type: [String, Number]
  },
  onAdd: Function,
  onChange: Function,
  onDragSort: Function,
  onRemove: Function
};

exports["default"] = props;
