// tdesign-vue-next/cjs/image/props.js
/**
 * tdesign v0.24.9
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  alt: {
    type: String,
    "default": ""
  },
  disabled: Boolean,
  error: {
    type: Object
  },
  fit: {
    type: String,
    "default": "fill",
    validator: function validator(val) {
      if (!val) return true;
      return ["contain", "cover", "fill", "none", "scale-down"].includes(val);
    }
  },
  gallery: Boolean,
  lazy: Boolean,
  loading: {
    type: Object
  },
  overlayContent: {
    type: Object
  },
  overlayTrigger: {
    type: String,
    "default": "always",
    validator: function validator(val) {
      if (!val) return true;
      return ["always", "hover"].includes(val);
    }
  },
  placeholder: {
    type: Object
  },
  position: {
    type: String,
    "default": "center"
  },
  shape: {
    type: String,
    "default": "square",
    validator: function validator(val) {
      if (!val) return true;
      return ["circle", "round", "square"].includes(val);
    }
  },
  src: {
    type: String,
    "default": ""
  },
  onError: Function,
  onLoad: Function
};

exports["default"] = props;
