// tdesign-vue-next/cjs/dialog/props.js
/**
 * tdesign v0.24.9
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  attach: {
    type: [String, Function],
    "default": ""
  },
  body: {
    type: [String, Function],
    "default": ""
  },
  cancelBtn: {
    type: [String, Object, Function],
    "default": ""
  },
  closeBtn: {
    type: [String, Boolean, Function],
    "default": true
  },
  closeOnEscKeydown: {
    type: Boolean,
    "default": void 0
  },
  closeOnOverlayClick: {
    type: Boolean,
    "default": void 0
  },
  confirmBtn: {
    type: [String, Object, Function],
    "default": ""
  },
  confirmOnEnter: Boolean,
  "default": {
    type: [String, Function]
  },
  destroyOnClose: Boolean,
  draggable: Boolean,
  footer: {
    type: [Boolean, Function],
    "default": true
  },
  header: {
    type: [String, Boolean, Function],
    "default": true
  },
  mode: {
    type: String,
    "default": "modal",
    validator: function validator(val) {
      if (!val) return true;
      return ["modal", "modeless", "normal", "full-screen"].includes(val);
    }
  },
  placement: {
    type: String,
    "default": "top",
    validator: function validator(val) {
      if (!val) return true;
      return ["top", "center"].includes(val);
    }
  },
  preventScrollThrough: {
    type: Boolean,
    "default": true
  },
  showInAttachedElement: Boolean,
  showOverlay: {
    type: Boolean,
    "default": true
  },
  theme: {
    type: String,
    "default": "default",
    validator: function validator(val) {
      if (!val) return true;
      return ["default", "info", "warning", "danger", "success"].includes(val);
    }
  },
  top: {
    type: [String, Number]
  },
  visible: Boolean,
  width: {
    type: [String, Number]
  },
  zIndex: {
    type: Number
  },
  onCancel: Function,
  onClose: Function,
  onCloseBtnClick: Function,
  onClosed: Function,
  onConfirm: Function,
  onEscKeydown: Function,
  onOpened: Function,
  onOverlayClick: Function
};

exports["default"] = props;
