// tdesign-vue-next/cjs/pagination/useMoreAction.js
/**
 * tdesign v0.24.9
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');

function useMoreAction(props, pageCount, innerCurrent) {
  var prevMore = vue.ref(false);
  var nextMore = vue.ref(false);
  var curPageLeftCount = vue.computed(function () {
    return Math.ceil((props.foldedMaxPageBtn - 1) / 2);
  });
  var curPageRightCount = vue.computed(function () {
    return Math.ceil((props.foldedMaxPageBtn - 1) / 2);
  });
  var isPrevMoreShow = vue.computed(function () {
    return 2 + curPageLeftCount.value < innerCurrent.value;
  });
  var isNextMoreShow = vue.computed(function () {
    return pageCount.value - 1 - curPageRightCount.value > innerCurrent.value;
  });
  return {
    prevMore: prevMore,
    nextMore: nextMore,
    curPageLeftCount: curPageLeftCount,
    curPageRightCount: curPageRightCount,
    isPrevMoreShow: isPrevMoreShow,
    isNextMoreShow: isNextMoreShow
  };
}

exports["default"] = useMoreAction;
