// tdesign-vue-next/cjs/guide/props.js
/**
 * tdesign v0.24.9
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  counter: {
    type: Function
  },
  current: {
    type: Number,
    "default": void 0
  },
  modelValue: {
    type: Number,
    "default": void 0
  },
  defaultCurrent: {
    type: Number
  },
  finishButtonProps: {
    type: Object,
    "default": {
      content: "\u5B8C\u6210",
      theme: "primary"
    }
  },
  hideCounter: Boolean,
  hidePrev: Boolean,
  hideSkip: Boolean,
  highlightPadding: {
    type: Number,
    "default": 8
  },
  mode: {
    type: String,
    "default": "popup",
    validator: function validator(val) {
      if (!val) return true;
      return ["popup", "dialog"].includes(val);
    }
  },
  nextButtonProps: {
    type: Object,
    "default": {
      content: "\u4E0B\u4E00\u6B65",
      theme: "primary"
    }
  },
  prevButtonProps: {
    type: Object,
    "default": {
      content: "\u4E0A\u4E00\u6B65",
      theme: "default"
    }
  },
  showOverlay: {
    type: Boolean,
    "default": true
  },
  skipButtonProps: {
    type: Object,
    "default": {
      content: "\u8DF3\u8FC7",
      theme: "default"
    }
  },
  steps: {
    type: Array
  },
  zIndex: {
    type: Number,
    "default": 999999
  },
  onChange: Function,
  onFinish: Function,
  onNextStepClick: Function,
  onPrevStepClick: Function,
  onSkip: Function
};

exports["default"] = props;
